### DISCUSSION TITLE
Complex control component pattern

### DISCUSSION THREAD

**Discussion by mortmoe** [react-hook-form/react-hook-form#2010] [General] [answered]
Just wondered if anyone have any feedback on this pattern for using complex input components with react-hook-form? Basically the idea is to have an internal state and useEffect to call onChange whenever the internal state changes. The component in the CodeSandbox is pretty simple and random, but in theory the internal value could be a lot more complex... ie. a geoJson object defining something a user has drawn on a map right?

https://codesandbox.io/s/react-hook-form-custom-input-pattern-lystm?file=/src/index.js

**Comment by bluebill1049** [comment, marked as answer]
honestly, i would build a controlled component and then wrapped with `Controller`, and let the `Controller` collect whatever data is necessary and with validation.
